Kernel/Bus/USB/UHCI/UHCIController.cpp

```diff
@@ -556,6 +600,31 @@ ErrorOr<void> UHCIController::spawn_port_process()
     return {};
 }
 
+ErrorOr<void> UHCIController::spawn_async_poll_process()
+{
+    LockRefPtr<Thread> async_poll_thread;
+    (void)Process::create_kernel_process(async_poll_thread, TRY(KString::try_create("UHCI Async Poll Task"sv)), [&] {
+        u16 poll_interval_ms = 1024;
+        for (;;) {
+            SpinlockLocker locker(m_async_lock);
+
+            for (auto& handle : m_active_async_transfers) {
+                poll_interval_ms = min(poll_interval_ms, handle.ms_poll_interval);
+                QueueHead* qh = handle.qh;
+                for (auto td = qh->get_first_td(); td != nullptr && !td->active(); td = td->next_td()) {
+                    if (td->next_td() == nullptr) { // Finished QH
+                        handle.transfer->invoke_async_callback();
+                        qh->reinitialize(); // Set the QH to be active again
+                    }
+                }
+            }
+            locker.unlock();
+            (void)Thread::current()->sleep(Time::from_milliseconds(poll_interval_ms));
```

Comment:
I think the only thing missing to make this "Semaphore at home" is a way to wake this thread up when appending to the queue. Just an unconditional `async_poll_thread->unblock()` from the "add transfer" method should do the trick. The no-args version, since that will call `Thread::unblock(u8 signal)` with a signal of 0, which will not dispatch any signals, just remove any thread blockers, namely the SleepBlocker from Thread::sleep().